### ISSUE TITLE
Restore-DBADatabase hangs if a partial uploaded BAK file exists

### ISSUE THREAD

**Issue by jeffincognito** [sqlcollaborative/dbatools#7921]
### Verified issue does not already exist?

Yes

### What error did you receive?

I'm working on a project to move a database into an azure environment daily for reporting. The file transmission happens through SFTP and as the file is uploaded, it's given a '.bak.part' file extension. 
This failed one day as the timing of the upload crossed the restore script and the process hung.  I was able to manually confirm that anytime the provided path is scanned and a file exists with the extension ".bak.part", the process will 
 stop and hang until manually killed. It's unclear if the issue is the extension or because the file has a lock on it.

### Steps to Reproduce

```powershell
# provide your command(s) executed pertaining to dbatools
# please include variable values (redacted or fake if needed) for reference
```
I was able to reproduce this by having 1 completed bak file in the restore path while uploading another through SFTP. I manually ran restore-dbadatabase pointed to the path of both files.

### Are you running the latest release?

Yes

### What PowerShell host was used when producing this error

Windows PowerShell (powershell.exe)

### PowerShell Host Version

5.1.17763.2183

### SQL Server Edition and Build number

13.0.5888.11

### .NET Framework Version

4.6.1

**Comment by andreasjordan** [comment]
As the files are not locally to the PowerShell session but locally to the SQL Server instance, we use `xp_dirtree` to get a list of files and pass them to `Read-DbaBackupHeader` to get the backup information.

So first thing that you could do to help us is to try only `Read-DbaBackupHeader` with a short list of files including the "part" file, to see if this hangs as well. The command uses runspaces to parallelize the processing and I don't know if it is possible to detect your situation. Let's ping @potatoqualitee to see if she has an idea.

**Comment by potatoqualitee** [comment]
I'm impressed that SMO can read a partially uploaded file! `bak.part` seems like an ideal filename to exclude in our process.

**Comment by andreasjordan** [comment]
Have a look at the linked pull request.

**Comment by wsmelton** [comment]
I would caution against filtering out explicit file extensions. SQL Server nor Azure have anything to do with those files being named `bak.part` (nothing documented that I could find).

SQL Server allows you to name any backup file what you want. I can create a backup file as `dont.deleteme` and it will try to restore it. Restore header is used for that purpose to validate it is a valid, formatted backup file.

The process hanging is a separate issue and could be any number of things in that processing. When you try to read a file that is locked by another process (here the uploading process) it can pause and wait for the other process to complete. 

There are ways to check whether a file can be touched or not but I don't think that is the fix here. If you are uploading your files via SFTP, you shouldn't call the Restore command until that is complete. Your own process should be the one that performs the verification whether the file has the expected extension. If we filter it out on our side then we can end up preventing a user from restoring a backup file they choose to name `.bak.part`.

**Comment by andreasjordan** [comment]
Thanks Shawn for your feedback, you have a very valid point here. 
So after that, I would vote for closing this issue as "works as designed". So the user would have to test the files for beeing all valid.

**Comment by Stuart-Moore** [comment]
Yes. By design restore looks at all files passed into it. Over the years I've come across a lot of stupid file extensions on backups, and even some without. 

Maybe a note/example in the help to show filtering the files passed into the command?

**Comment by andreasjordan** [comment]
I will change this from a bug to a documentation issue...

**Comment by andreasjordan** [comment]
@claude pleasse suggest a suitable documentation to address this issue.
